pipeline/setup-runner.py

```diff
@@ -0,0 +1,347 @@
+    },
+}
+
+
+def run(cmd, check=True):
+    print(f"Running: {cmd if isinstance(cmd, str) else ' '.join(cmd)}")
+    result = subprocess.run(cmd, check=False)
+    if check and result.returncode != 0:
+        sys.exit(result.returncode)
+    return result
+
+
+def download_from_s3(s3_path, local_path):
+    bucket = os.environ.get("INSTALLER_BUCKET")
+    if not bucket:
+        print("ERROR: INSTALLER_BUCKET not set")
+        sys.exit(1)
+
+    expected_bucket_owner = os.environ.get("INSTALLER_BUCKET_EXPECTED_OWNER")
+    if not expected_bucket_owner:
+        raise ValueError("INSTALLER_BUCKET_EXPECTED_OWNER environment variable is required")
+    if not (expected_bucket_owner.isdigit() and len(expected_bucket_owner) == 12):
+        raise ValueError("INSTALLER_BUCKET_EXPECTED_OWNER must be a 12-digit AWS Account ID")
+
+    config = Config(read_timeout=300, connect_timeout=60, retries={"max_attempts": 2})
+
+    s3 = boto3.client("s3", config=config)
+
+    print(f"Downloading s3://{bucket}/{s3_path} to {local_path}")
+
+    s3.download_file(
+        bucket, s3_path, str(local_path), ExtraArgs={"ExpectedBucketOwner": expected_bucket_owner}
+    )
+
+
+def verify_checksum(file_path, expected_checksum):
+    """Verify SHA256 checksum of downloaded file."""
+    print(f"Verifying checksum for {file_path}...")
+    sha256 = hashlib.sha256()
+    with open(file_path, "rb") as f:
+        for chunk in iter(lambda: f.read(8192), b""):
+            sha256.update(chunk)
+
+    actual = sha256.hexdigest()
+    if actual != expected_checksum:
+        print("ERROR: Checksum mismatch!")
+        print(f"  Expected: {expected_checksum}")
+        print(f"  Actual:   {actual}")
+        sys.exit(1)
+
+    print("OK Checksum verified")
+    return True
+
+
+def setup_linux():
+    # Install bc (required by Houdini installer)
+    pkg_mgr = (
+        "dnf"
+        if subprocess.run(["command", "-v", "dnf"], capture_output=True, check=False).returncode
+        == 0
+        else "yum"
+    )
+    run([pkg_mgr, "install", "-y", "bc"])
+
+    for version in HOUDINI_VERSIONS:
+        major_minor = ".".join(version.split(".")[:2])
+        houdini_dir = Path(f"/opt/hfs{version}")
+        houdini_marker = houdini_dir / ".installed"
+
+        if houdini_marker.exists():
+            print(f"Houdini {version} already installed")
+            continue
+
+        lock_file = Path(f"/tmp/houdini-{version}.lock")
+        if lock_file.exists():
+            print(f"Waiting for concurrent Houdini {version} install...")
+            for _ in range(120):
+                time.sleep(1)
+                if houdini_marker.exists():
+                    break
+            continue
+
+        lock_file.touch()
+        try:
+            print(f"Installing Houdini {version}...")
+            # Houdini 19.5 uses gcc9.3, others use gcc11.2
+            gcc_version = "gcc9.3" if version.startswith("19.5") else "gcc11.2"
```

Comment:
For more streamlined maintainability, I wonder if this could be maintained in a single data structure at the top of the file. What do you think?